Fixes #25917: Add a dedicated Readme for rudder-package (#6032)
Normation/rudder: relay/sources/rudder-pkg/Readme.md

> 

Reviewer: the convention is `README.md`, let's stick to it

# Rpkg spec

> A rpkg is an `ar` archive containing with the `.rpkg` extension containing:

Reviewer: ```suggestion
A rpkg is an `ar` archive with the `.rpkg` extension containing:
```

> * a metadata in **Json**

Reviewer: ```suggestion
* a metadata in **JSON**
```

  "depends": {
    "binary": [ "zip" ]
    "apt": [ "apache2" ],
    "rpm": [ ],
    "python": []
  },
  "content": {
    "files.txz": "/opt/rudder/share",
    "var_rudder.txz": "/var/rudder"
  }
}
```
| **Field**          | **Type**               | **Mandatory** | **Details**                                                                                     |
|---------------------|------------------------|---------------|-------------------------------------------------------------------------------------------------|
| `type`             | String                | Yes           | Specifies the type of the rpkg. Only the value `"plugin"` is currently supported.                               |
| `name`             | String                | Yes           | The name of the plugin.                                    |
| `version`          | String                | Yes           | The version of the compatible webapp and the version of the plugin concatenated by a `-`.                                 |
| `description`      | String                | No            | A brief description of the plugin. This can be empty or omitted.                              |

> | `build-date`       | String (ISO 8601)     | Yes           | The date and time when the plugin was built, in ISO 8601 format.                              |

Reviewer: ```suggestion
| `build-date`       | String (RFC 3339)     | Yes           | The date and time when the plugin was built, in RFC 3339 format.                              |
```

Reviewer: it is mostly a subset

| `build-commit`     | String (SHA1 hash)    | Yes           | The Git commit hash associated with the build.                                                |
| `jar-files`        | Array of Strings      | Yes           | A list of jar file names included in the plugin. An empty list if no jar files are needed.                                          |

> | `depends`          | Object                | No           | Lists dependencies required for the plugin to function. Each key represents a type of dependency (e.g., `binary`, `apt`, `rpm`, `python`) and maps to an array of specific dependencies. |

Reviewer: ```suggestion
| `depends`          | Object                | No           | Lists dependencies required for the plugin to function. Each key represents a type of dependency (`binary`, `apt`, `rpm` or `python`) and maps to an array of specific dependencies. |
```

Reviewer: the list is set in the spec as it requires a specific implementation

| `content`          | Object                | Yes           | Maps file archives to their installation paths within the system.                             |

### Subfields of `depends`
| **Field** | **Type**          | **Mandatory** | **Details**                                                                  |
|-----------|-------------------|---------------|------------------------------------------------------------------------------|

> | `binary`  | Array of Strings | No            | Dependencies that are binary packages.                                        |

Reviewer: ```suggestion
| `binary`  | Array of Strings | No            | Dependencies that are binary packages and have to be available in `PATH`.                                  |
```

> | `apt`     | Array of Strings | No            | Dependencies for Debian-based systems.                                        |

Reviewer: ```suggestion
| `apt`     | Array of Strings | No            | Dependencies for APT-based systems.                                        |
```

Reviewer: for consistence

| 8.0.2~beta1-2.9           | 8.0.2~beta1              | ✅             |
| 8.0.2~alpha1.2-2.9        | 8.0.2\~alpha2~git12345    | ❌             |

## Package scripts

Package scripts are executables that are run at specific timing during the installation/uninstallation process.
They are expected to return 0, otherwise the installation/uninstallation will be aborted.

They **must** be packaged inside the `scripts.txz` archive in the rpkg. And they will always be unpacked on the Rudder server under `/var/rudder/packages/rudder-plugin-<plugin name from metadata>`.

## Plugin content

A plugin can contain multiple archives that will be unpacked on different locations on the system at install time.
Each file deployed this way will be removed when the plugin will be uninstalled.

Each content archive must be listed in the metadata under the `contents` field, using the archive name as key, and the target installation path as value.

```json
  "content": {
    "files.txz": "/opt/rudder/share",
    "var_rudder.txz": "/var/rudder"
  }
```

## RPKG installation flowchart

A failure in any of the steps in the schema below will abort the plugin installation without cleaning.

> [![](https://mermaid.ink/img/pako:eNp1U0tz2jAQ_isanSHExpTEh840QBKSaQ-0PrR2D8JaQMWWNJJMQoD_XlmSmdCHDiut9vl9Wh1wKSjgFK8q8VJuiDLo27TgyC7dLNeKyA1ayO0aMa4NqSpimODe_ilfAKHIbADVYAglhvz0lrt8soFyi15gSaREpailjVuyipl9cJkEFwoSOAVeMtDBND3MNWIGEY4aaTugcPKGWT57NYqUBqm2I0nKLVkD0qVi0nTR9_mi4UgqaBsOtmB6uIwvBTfAO-NjnkkLARyeABYoklWzZhy14JZEQ3Ce-yJCm7-rPOUzTpYVoF9EoRWrwIJZWTAd8mdLm81uifb0hGvLwh-0Zzy0gURFQaEdKH0mP4vyKdOuTtvwf2plsS1Wi513UmCfz2LyjhXTrWLBtbZ_48yG7-MpUFa6DMF75Ruzp8GOqIFqqFUH4V26B8mS9zm6SCVqn_OyoGMhjBfq9z-iOycnToa5nDrl-EUcZxcX30EfsyiMiru6d_LByUcn504-OflcnLl0ehb7bei3xG22BO7hGlRNGLXf5NCGFNi2XkOBU3ukRG0LXPCT9SONEV_3vMSpUQ30sBLNeoPTFam01Ro3YFNG7PvW51tJ-A8h6i7EcmyE-uw_pfubzgWnB_yK02QYXyXJTZzE0c141MN7nPavr67bNYqj29toPIqi0YdxMh6eevjN5Y1OvwGl5T4S?type=png)](https://mermaid.live/edit#pako:eNp1U0tz2jAQ_isanSHExpTEh840QBKSaQ-0PrR2D8JaQMWWNJJMQoD_XlmSmdCHDiut9vl9Wh1wKSjgFK8q8VJuiDLo27TgyC7dLNeKyA1ayO0aMa4NqSpimODe_ilfAKHIbADVYAglhvz0lrt8soFyi15gSaREpailjVuyipl9cJkEFwoSOAVeMtDBND3MNWIGEY4aaTugcPKGWT57NYqUBqm2I0nKLVkD0qVi0nTR9_mi4UgqaBsOtmB6uIwvBTfAO-NjnkkLARyeABYoklWzZhy14JZEQ3Ce-yJCm7-rPOUzTpYVoF9EoRWrwIJZWTAd8mdLm81uifb0hGvLwh-0Zzy0gURFQaEdKH0mP4vyKdOuTtvwf2plsS1Wi513UmCfz2LyjhXTrWLBtbZ_48yG7-MpUFa6DMF75Ruzp8GOqIFqqFUH4V26B8mS9zm6SCVqn_OyoGMhjBfq9z-iOycnToa5nDrl-EUcZxcX30EfsyiMiru6d_LByUcn504-OflcnLl0ehb7bei3xG22BO7hGlRNGLXf5NCGFNi2XkOBU3ukRG0LXPCT9SONEV_3vMSpUQ30sBLNeoPTFam01Ro3YFNG7PvW51tJ-A8h6i7EcmyE-uw_pfubzgWnB_yK02QYXyXJTZzE0c141MN7nPavr67bNYqj29toPIqi0YdxMh6eevjN5Y1OvwGl5T4S)

Reviewer: GitHub can render mermaid directly https://github.blog/developer-skills/github/include-diagrams-markdown-files-mermaid/

Reviewer: (and most graphical text editors too)

> ## Guidelines to write a Plugin

Reviewer: ```suggestion
## Guidelines to write a plugin
```
